flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/queue/UnorderedStreamElementQueue.java

```diff
@@ -19,192 +19,146 @@
 package org.apache.flink.streaming.api.operators.async.queue;
 
 import org.apache.flink.annotation.Internal;
-import org.apache.flink.streaming.api.operators.async.OperatorActions;
+import org.apache.flink.streaming.api.functions.async.ResultFuture;
+import org.apache.flink.streaming.api.operators.TimestampedCollector;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamElement;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
 import org.apache.flink.util.Preconditions;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.ArrayDeque;
-import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Deque;
 import java.util.HashSet;
-import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Queue;
 import java.util.Set;
-import java.util.concurrent.Executor;
-import java.util.concurrent.locks.Condition;
-import java.util.concurrent.locks.ReentrantLock;
 
 /**
- * Unordered implementation of the {@link StreamElementQueue}. The unordered stream element queue
- * emits asynchronous results as soon as they are completed. Additionally it maintains the
- * watermark-stream record order. This means that no stream record can be overtaken by a watermark
- * and no watermark can overtake a stream record. However, stream records falling in the same
- * segment between two watermarks can overtake each other (their emission order is not guaranteed).
+ * Unordered implementation of the {@link StreamElementQueue}. The unordered stream element queue provides
+ * asynchronous results as soon as they are completed. Additionally it maintains the watermark-stream record order.
```

Comment:
The previous descriptions for order/unorder issues are not very clearly in my thought.  Especially for `no stream record can be overtaken by a watermark and no watermark can overtake a stream record`, the sayings before/after `and` are the same.
I think it is better to involve in the stage concept for better understanding. If I understood correctly, elements in one stage are kept the order, and elements in different stages are not in order which means overtaking each other.

Reply:
It's actually the other way around: only elements within a stage can overtake each other. I will add a description to Stage and rewrite this description.

Reply:
Yes. Elements within a stage could overtake each other. One more thing for further confirmation: how to guarantee that two stages can not overtake each other? 

e.g. assuming we have four stages : {a, b, c} {watermark1} {d, e, f} {watermark2}, and the first stage is the {a, b, c}. If the user function completes the element `e` in third stage firstly, that means the collection of `e` would be emitted to output before the first stage? Or I missed something else.